### PR TITLE
encryption: add oob migration for user external accounts

### PR DESCRIPTION
**Pull request by asdine**
This PR adds a background job that encrypts user external accounts at startup.
It uses the new out of band migration package https://github.com/sourcegraph/sourcegraph/blob/main/doc/dev/background-information/oobmigrations.md

### CODE REVIEW

File: internal/database/external_accounts.go [hunk 1]
```diff
@@ -502,3 +503,153 @@ type MockExternalAccounts struct {
...
+	return &ExternalAccountsMigrator{store: store, BatchSize: 50}
+}
+
+func NewExternalAccountsMigratorWithDB(db dbutil.DB) *ExternalAccountsMigrator {
+	return NewExternalAccountsMigrator(basestore.NewWithDB(db, sql.TxOptions{}))
+}
+
+// ID of the migration row in the out_of_band_migrations table.
+func (m *ExternalAccountsMigrator) ID() int {
+	return 4
```
Comment (unknwon) [inline internal/database/external_accounts.go:530]: This feels a bit hacky TBH... we can probably have a dedicated file that lists out all IDs.

File: internal/database/external_accounts.go [hunk 2]
```diff
@@ -502,3 +503,153 @@ type MockExternalAccounts struct {
...
 	TouchLastValid       func(ctx context.Context, id int32) error
 }
+
+// ExternalServiceConfigMigrator is a background job that encrypts
+// external services config on startup.
+// It periodically waits until a keyring is configured to determine
+// how many services it must migrate.
+// Scheduling and progress report is deleguated to the out of band
+// migration package.
+// The migration is non destructive and can be reverted.
```
Comment (unknwon) [inline internal/database/external_accounts.go:513]: Is it still true if key changed when running `Down`?

**Review by unknwon — COMMENTED** [review]
I think `account_data` is not worthing encrypting. It contains no secrets.

**Comment by asdine** [comment]
> I think account_data is not worthing encrypting. It contains no secrets.

@unknwon Noted, I will only encrypt `auth_data` then. https://github.com/sourcegraph/sourcegraph/pull/19270 has already landed but it's safe to revert just the encryption of `auth_data` as the external account encryption key hasn't yet been activated.

File: internal/database/external_accounts.go (hunk 2 above)
Comment (asdine) [reply internal/database/external_accounts.go]: No, we have yet to define the behavior for key rotation. It's only true for the same key

File: internal/database/external_accounts.go (hunk 1 above)
Comment (asdine) [reply internal/database/external_accounts.go]: For now, I added a link to the migration file, but yes we will need to define a good way to define all of these ids.

File: internal/database/oob_migrate.go [hunk 3]
```diff
@@ -178,3 +178,154 @@ func (m *ExternalServiceConfigMigrator) listConfigsForUpdate(ctx context.Context
...
+			decrypted, err := key.Decrypt(ctx, encAuthData)
+			if err != nil {
+				return err
+			}
+			if decrypted.Secret() != string(*acc.AuthData) {
+				return errors.New("invalid encryption round-trip")
+			}
+		}
+
+		if acc.Data != nil {
```
Comment (unknwon) [inline internal/database/oob_migrate.go:269]: I noticed you have removed encryption code in application code for `acc.Data`, should we also remove this if block entirely?

Comment (asdine) [reply internal/database/oob_migrate.go:269]: Good catch!

**Review by unknwon — APPROVED** [review]

**Pull request merged**